presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java

```diff
@@ -0,0 +1,186 @@
+                }
+            });
+        }
+    }
+
+    private boolean matchExceptionByErrorMessage(Exception exception, String errorMessage)
+    {
+        return getCausalChain(exception).stream()
+                .anyMatch(e -> e.getMessage().equals(errorMessage));
+    }
+
+    private void assertAsyncTaskExecution(int maxAttempts, List<AttemptStatus> scheduledAttempts, AttemptStatus expectedFinalStatus)
+    {
+        RetryDriver driver = RetryDriver.retry(retryExecutor)
+                .maxAttempts(maxAttempts)
+                .exponentialBackoff(minDuration, maxDuration, maxRetryTime, backOffScaleFactor);
+        MockAsyncTask task = new MockAsyncTask(scheduledAttempts, taskExecutor);
+
+        try {
+            // execute and wait on the future
+            driver.runAsync("task", task).get();
+            assertLessThanOrEqual(task.getNextAttempt(), maxAttempts, "Final attempts should be equal or less than maxAttempt. Task: " + task);
+            if (expectedFinalStatus != AttemptStatus.SUCCESS) {
+                fail("Task is expected to fail but actually succeeded. Task: " + task);
+            }
+        }
+        catch (Exception ex) {
+            assertEquals(task.getNextAttempt(), maxAttempts, "Failure must happen on max attempt. Task: " + task);
+            if (expectedFinalStatus == AttemptStatus.SUCCESS) {
+                fail("Task is expected to fail but actually succeeded. Task: " + task);
+            }
+            else if (expectedFinalStatus.equals(AttemptStatus.ASYNC_FAILURE) && !matchExceptionByErrorMessage(ex, ASYNC_FAILURE_MESSAGE)) {
+                fail("Task is expected to fail due to async error. Actual error: " + ex);
+            }
+            else if (expectedFinalStatus.equals(AttemptStatus.SYNC_FAILURE) && !matchExceptionByErrorMessage(ex, SYNC_FAILURE_MESSAGE)) {
+                fail("Task is expected to fail due to sync error. Actual error: " + ex);
+            }
+        }
+    }
+
+    @Test
+    void testRetryDriver()
+    {
+        List<AttemptStatus> twoAttempts = ImmutableList.of(AttemptStatus.ASYNC_FAILURE, AttemptStatus.SUCCESS);
+        for (int i = 1; i <= twoAttempts.size(); ++i) {
+            assertAsyncTaskExecution(i, twoAttempts, twoAttempts.get(i - 1));
+        }
+        // one more max attempt
+        assertAsyncTaskExecution(twoAttempts.size() + 1, twoAttempts, twoAttempts.get(twoAttempts.size() - 1));
+
+        List<AttemptStatus> fiveAttempts = ImmutableList.of(AttemptStatus.SYNC_FAILURE, AttemptStatus.ASYNC_FAILURE, AttemptStatus.SYNC_FAILURE, AttemptStatus.ASYNC_FAILURE, AttemptStatus.SUCCESS);
+        for (int i = 1; i <= fiveAttempts.size(); ++i) {
+            assertAsyncTaskExecution(i, fiveAttempts, fiveAttempts.get(i - 1));
+        }
+        // one more max attempt
+        assertAsyncTaskExecution(fiveAttempts.size() + 1, fiveAttempts, fiveAttempts.get(fiveAttempts.size() - 1));
+    }
```

Comment:
It looks like you have several cases in one test. Typically, you need to have one case per method.

Reply:
Refactored test cases.